src/symbolic_dimensions.jl

```diff
@@ -95,14 +99,118 @@ end
 expand_units(q::QuantityArray) = expand_units.(q)
 
 
-Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(data(d)))
-Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions) = data(l) == data(r)
-Base.iszero(d::SymbolicDimensions) = iszero(data(d))
-Base.:*(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) + data(r))
-Base.:/(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) - data(r))
-Base.inv(d::SymbolicDimensions) = SymbolicDimensions(-data(d))
-Base.:^(l::SymbolicDimensions{R}, r::Integer) where {R} = SymbolicDimensions(data(l) * r)
-Base.:^(l::SymbolicDimensions{R}, r::Number) where {R} = SymbolicDimensions(data(l) * tryrationalize(R, r))
+Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(getfield(d, :nzdims)), copy(getfield(d, :nzvals)))
+function Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions)
+    nzdimsl = getfield(l, :nzdims)
+    nzvalsl = getfield(l, :nzvals)
+    nzdimsr = getfield(r, :nzdims)
+    nzvalsr = getfield(r, :nzvals)
+    nl = length(nzdimsl)
+    nr = length(nzdimsr)
+    il = ir = 1
+    while il <= nl && ir <= nr
+        diml = nzdimsl[il]
+        dimr = nzdimsr[ir]
+        if diml == dimr
+            if nzvalsl[il] != nzvalsr[ir]
+                return false
+            end
+            il += 1
+            ir += 1
+        elseif diml < dimr
+            if !iszero(nzvalsl[il])
+                return false
+            end
+            il += 1
+        else
+            if !iszero(nzvalsr[ir])
+                return false
+            end
+            ir += 1
+        end
+    end
+
+    while il <= nl
+        if !iszero(nzvalsl[il])
+            return false
+        end
+        il += 1
+    end
+
+    while ir <= nr
+        if !iszero(nzvalsr[ir])
+            return false
+        end
+        ir += 1
+    end
+
+    return true
+end
+Base.iszero(d::SymbolicDimensions) = iszero(getfield(d, :nzvals))
+Base.:*(l::SymbolicDimensions, r::SymbolicDimensions) = _combine_vals(+, l, r)
+Base.:/(l::SymbolicDimensions, r::SymbolicDimensions) = _combine_vals(-, l, r)
+function _combine_vals(op::O, l::SymbolicDimensions{L}, r::SymbolicDimensions{R}) where {O,L,R}
```

Comment:
I think it might be better to use multiple dispatch on `map_dimensions` here instead of defining a new function name. Then some operators come for free like `*` and `/`.

Reply:
Also the single-operator `map_dimensions` should be defined as well, which would give us `inv` for free.

Reply:
I just defined whatever is currently defined on the master branch. I guess additional improvements might be better left for a separate PR?

Reply:
Oh, sorry, what I’m trying to say is that we only need to change the function name `_combine_vals` to `map_dimensions`. Rather than defining a new function name. (This signature is the same one used for `map_dimensions`)

Then the lines for `*` and `/` can be deleted as well.

Reply:
Ah, I see 👍

Reply:
Done. BTW what's the reason for a separate `map_dimensions` function instead of just overloading `map`?